drivers/flash/flash_spi.c

```diff
@@ -0,0 +1,382 @@
+    } while (retval & FLASH_FLAG_WIP);
+
+    return response;
+}
+
+static flash_status_t spi_flash_write_enable(void) {
+
+    bool res = spi_flash_start();
+    if (!res) {
+        dprint("Failed to start SPI! [spi flash write enable]\n");
+        return FLASH_STATUS_ERROR;
+    }
+
+    spi_write(FLASH_CMD_WREN);
+
+    spi_stop();
+
+    return FLASH_STATUS_SUCCESS;
+}
+
+static flash_status_t spi_flash_write_disable(void) {
+
+    bool res = spi_flash_start();
+    if (!res) {
+        dprint("Failed to start SPI! [spi flash write disable]\n");
+        return FLASH_STATUS_ERROR;
+    }
+
+    spi_write(FLASH_CMD_WRDI);
+
+    spi_stop();
+
+    return FLASH_STATUS_SUCCESS;
+}
+
+/* This function is used for read transfer, write transfer and erase transfer. */
+static flash_status_t spi_flash_transmit(uint8_t cmd, uintptr_t addr, uint8_t *data, size_t len) {
+    flash_status_t response = FLASH_STATUS_SUCCESS;
+    uint8_t buffer[EXTERNAL_FLASH_ADDRESS_SIZE + 1];
+
+    buffer[0] = cmd;
+    for (int i = 0; i < EXTERNAL_FLASH_ADDRESS_SIZE; ++i) {
+        buffer[EXTERNAL_FLASH_ADDRESS_SIZE - i] = addr & 0xFF;
+        addr >>= 8;
+    }
+
+    bool res = spi_flash_start();
+    if (!res) {
+        dprint("Failed to start SPI! [spi flash transmit]\n");
+        return FLASH_STATUS_ERROR;
+    }
+
+    response = spi_transmit(buffer, sizeof(buffer));
+
+    if ((!response) && (data != NULL)) {
+        switch (cmd) {
+          case FLASH_CMD_READ:
+              response = spi_receive(data, len);
+              break;
+          case FLASH_CMD_PP:
+              response = spi_transmit(data, len);
+              break;
+          default:
+              response = FLASH_STATUS_ERROR;
+              break;
+        }
+    }
+
+    spi_stop();
+
+    return response;
+}
+
+void flash_init(void) {
+    spi_init();
+}
+
+flash_status_t flash_erase_chip(void) {
+    flash_status_t response = FLASH_STATUS_SUCCESS;
+
+    /* Wait for the write-in-progress bit to be cleared. */
+    response= spi_flash_wait_while_busy();
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to check WIP flag! [spi flash erase chip]\n");
+        return response;
+    }
+
+    /* Enable writes. */
+    response = spi_flash_write_enable();
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to write-enable! [spi flash erase chip]\n");
+        return response;
+    }
+
+    /* Erase Chip. */
+    bool res = spi_flash_start();
+    if (!res) {
+        dprint("Failed to start SPI! [spi flash erase chip]\n");
+        return FLASH_STATUS_ERROR;
+    }
+    spi_write(FLASH_CMD_CE);
+    spi_stop();
+
+    /* Wait for the write-in-progress bit to be cleared.*/
+    response= spi_flash_wait_while_busy();
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to check WIP flag! [spi flash erase chip]\n");
+        return response;
+    }
+
+    return response;
+}
+
+flash_status_t flash_erase_sector(const void *addr) {
```

Comment:
```suggestion
flash_status_t flash_erase_sector(uint32_t addr) {
```
Please also swap other usages of `target_addr` to just `addr`.